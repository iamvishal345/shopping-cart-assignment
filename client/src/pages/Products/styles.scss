@import "assets/scss/_breakpoints.scss";

.products-container {
  display: flex;
  @include xm {
    flex-direction: column;
  }
  .products-list {
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    margin: 8px 0 1rem 8px;
    grid-column-gap: 4px;
    @include md {
      grid-template-columns: repeat(2, 1fr);
    }
    @include xm {
      grid-template-columns: repeat(1, 1fr);
    }
  }
}
