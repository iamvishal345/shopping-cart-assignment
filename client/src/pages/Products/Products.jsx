import React, { useContext, useEffect, useRef, useState } from "react";
import "./styles.scss";
import { Sidebar } from "components/Atoms/Sidebar/Sidebar";
import { ProductCard } from "components/Atoms/ProductCard/ProductCard";
import { CartContext } from "context/CartContext";
import { Head } from "Utils/Head";

const Products = () => {
  const products = useRef();
  const [categories, setCategories] = useState([]);
  const [filteredProducts, setFilteredProducts] = useState([]);
  const addToCart = useContext(CartContext).addToCart;
  useEffect(() => {
    fetch("/api/categories")
      .then((res) => {
        return res.json();
      })
      .then((res) => {
        setCategories(res);
      })
      .catch((err) => {
        console.error(err);
      });
    fetch("/api/products")
      .then((res) => {
        return res.json();
      })
      .then((res) => {
        products.current = res;
        setFilteredProducts(res);
      })
      .catch((err) => {
        console.error(err);
      });
  }, []);

  const handleBuyNow = (product) => {
    addToCart(product);
  };
  const selectedCallback = (menuId) => {
    if (menuId === "all") setFilteredProducts([...products.current]);
    else
      setFilteredProducts([
        ...products.current.filter((prod) => prod.category === menuId),
      ]);
  };
  return (
    <React.Fragment>
      <Head
        title="Sabka Bazaar - Available Products"
        description="Fresh fruits and vegetables and genuine cosmetics and Baby care products"
      />
      <div className="products-container">
        {categories.length > 0 && (
          <Sidebar
            menuList={categories.sort((a, b) => a.order - b.order)}
            selectedCallback={selectedCallback}
          />
        )}
        <div className="products-list">
          {filteredProducts.map((product) => (
            <ProductCard
              key={product.id}
              product={product}
              handlePrimaryAction={handleBuyNow}
            />
          ))}
        </div>
      </div>
    </React.Fragment>
  );
};

export default Products;
