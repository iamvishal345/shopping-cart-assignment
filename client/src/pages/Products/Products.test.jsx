import React from "react";
import { act } from "react-dom/test-utils";
import { mount } from "enzyme";
import Products from "./Products";
import fetchMock from "fetch-mock";
// import { ProductCard } from "components/Atoms/ProductCard/ProductCard";
jest.mock("../../Utils/Head.jsx", () => ({
  Head: "div",
}));

const mockAddToCart = jest.fn();
jest.mock("react", () => {
  const ActualReact = jest.requireActual("react");
  return {
    ...ActualReact,
    useContext: () => ({
      addToCart: mockAddToCart,
    }),
  };
});
const categories = [
  {
    name: "Beverages",
    key: "beverages",
    description:
      "Our beverage department will ensure your fridge is always fully stocked. Shop for soda, juice, beer and more. ",
    enabled: true,
    order: 3,
    imageUrl: "/static/images/category/beverages.png",
    id: "5b675e5e5936635728f9fc30",
  },
  {
    name: "Bakery Cakes and Dairy",
    key: "bakery-cakes-dairy",
    description:
      "The best cupcakes, cookies, cakes, pies, cheesecakes, fresh bread, biscotti, muffins, bagels, fresh coffee, milk and more.",
    enabled: true,
    order: 2,
    imageUrl: "/static/images/category/bakery.png",
    id: "5b6899123d1a866534f516de",
  },
  {
    name: "Beauty and Hygiene",
    key: "beauty-hygiene",
    description:
      "Buy beauty and personal care products online in India at best prices.",
    enabled: true,
    order: 4,
    imageUrl: "/static/images/category/beauty.png",
    id: "5b68994e3d1a866534f516df",
  },
];

const mainProducts = [
  {
    name: "Fresho Kiwi - Green, 3 pcs",
    imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
    description:
      "Kiwis are oval shaped with a brownish outer skin. The flesh is bright green and juicy with tiny, edible black seeds.",
    price: 87,
    stock: 50,
    category: "5b6899953d1a866534f516e2",
    sku: "fnw-kiwi-3",
    id: "5b6c6a7f01a7c38429530883",
  },
  {
    name: "Apple - Washington, Regular, 4 pcs",
    imageURL: "/static/images/products/fruit-n-veg/apple.jpg",
    description:
      "The bright red coloured and heart shaped Washington apples are crunchy, juicy and slightly sweet. Washington Apples are a natural source of fibre and are fat free.",
    price: 187,
    stock: 50,
    category: "5b6899953d1a866534f516e2",
    sku: "fnw-apple-4",
    id: "5b6c6aeb01a7c38429530884",
  },
  {
    name: "Fresho Pomegrante Peeled, 500 gm ",
    imageURL: "/static/images/products/fruit-n-veg/pomegrante.jpg",
    description:
      "Pomegranate variety has a glossy, vibrant and soft ruby-red skin. The inside of the fruit is divided into compartments by thin white membranes.",
    price: 88,
    stock: 50,
    category: "5b6899953d1a866534f516e2",
    sku: "fnw-pomegranate-500",
    id: "5b6c6b7001a7c38429530885",
  },
];

fetchMock.get("end:/api/categories", categories);
fetchMock.get("end:/api/products", mainProducts);
let wrapped;
act(() => {
  wrapped = mount(<Products />);
});
describe("Products", () => {
  it("should match the snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });
});
