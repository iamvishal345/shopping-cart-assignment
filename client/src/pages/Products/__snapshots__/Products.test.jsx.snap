// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products should match the snapshot 1`] = `
<Products>
  <div
    description="Fresh fruits and vegetables and genuine cosmetics and Baby care products"
    title="Sabka Bazaar - Available Products"
  />
  <div
    className="products-container"
  >
    <div
      className="products-list"
    />
  </div>
</Products>
`;
